packages/rol/adapters/thyra/src/function/ROL_ThyraProductME_Constraint_SimOpt.hpp

```diff
@@ -54,15 +54,15 @@
 #include "Tpetra_CrsMatrix.hpp"
 #include "Teuchos_VerbosityLevel.hpp"
 
-using namespace ROL;
+namespace ROL {
 
 //! \brief ROL interface wrapper for Sacado SimOpt Constraint
 template<class Real>
 class ThyraProductME_Constraint_SimOpt : public Constraint_SimOpt<Real> {
 
 public:
 
-  ThyraProductME_Constraint_SimOpt(Thyra::ModelEvaluatorDefaultBase<double>& thyra_model_, int g_index_, const std::vector<int>& p_indices_,
+  ThyraProductME_Constraint_SimOpt(const Thyra::ModelEvaluator<double>& thyra_model_, int g_index_, const std::vector<int>& p_indices_,
```

Comment:
why did you have to change this?

Reply:
It was required to compile the `evalConvergedModelResponsesAndSensitivities` function.

Not doing those modifications leads to the following errors:

```
[ 85%] Built target ROL_test_compatibility_stacktrace
[ 85%] Building CXX object packages/piro/src/CMakeFiles/piro.dir/Piro_NOXSolver.cpp.o
In file included from /home/kim/dev/Trilinos_all/Trilinos/packages/piro/src/Piro_SteadyStateSolver.hpp:159:0,
                 from /home/kim/dev/Trilinos_all/Trilinos/packages/piro/src/Piro_NOXSolver.hpp:46,
                 from /home/kim/dev/Trilinos_all/Trilinos/packages/piro/src/Piro_NOXSolver_Def.hpp:46,
                 from /home/kim/dev/Trilinos_all/Trilinos/packages/piro/src/Piro_NOXSolver.cpp:43:
/home/kim/dev/Trilinos_all/Trilinos/packages/piro/src/Piro_SteadyStateSolver_Def.hpp: In instantiation of ‘void Piro::SteadyStateSolver<Scalar>::evalConvergedModelResponsesAndSensitivities(const Thyra::ModelEvaluatorBase::InArgs<Scalar>&, const Thyra::ModelEvaluatorBase::OutArgs<Scalar>&) const [with Scalar = double]’:
/home/kim/dev/Trilinos_all/Trilinos/packages/piro/src/Piro_NOXSolver_Def.hpp:204:3:   required from ‘void Piro::NOXSolver<Scalar>::evalModelImpl(const Thyra::ModelEvaluatorBase::InArgs<Scalar>&, const Thyra::ModelEvaluatorBase::OutArgs<Scalar>&) const [with Scalar = double]’
/home/kim/dev/Trilinos_all/Trilinos/packages/piro/src/Piro_NOXSolver.cpp:50:16:   required from here
/home/kim/dev/Trilinos_all/Trilinos/packages/piro/src/Piro_SteadyStateSolver_Def.hpp:431:52: error: no matching function for call to ‘ROL::ThyraProductME_Objective_SimOpt<double>::ThyraProductME_Objective_SimOpt(const Thyra::ModelEvaluator<double>&, int&, std::vector<int, std::allocator<int> >&, Teuchos::RCP<Teuchos::ParameterList>&, Teuchos::EVerbosityLevel)’
       ROL::ThyraProductME_Objective_SimOpt<Scalar> obj(*model_, i, p_indices, opt_paramList, Teuchos::VERB_NONE);
                                                    ^~~
In file included from /home/kim/dev/Trilinos_all/Trilinos/packages/piro/src/Piro_SteadyStateSolver_Def.hpp:79:0,
                 from /home/kim/dev/Trilinos_all/Trilinos/packages/piro/src/Piro_SteadyStateSolver.hpp:159,
                 from /home/kim/dev/Trilinos_all/Trilinos/packages/piro/src/Piro_NOXSolver.hpp:46,
                 from /home/kim/dev/Trilinos_all/Trilinos/packages/piro/src/Piro_NOXSolver_Def.hpp:46,
                 from /home/kim/dev/Trilinos_all/Trilinos/packages/piro/src/Piro_NOXSolver.cpp:43:
/home/kim/dev/Trilinos_all/Trilinos/packages/piro/src/../../rol/adapters/thyra/src/function/ROL_ThyraProductME_Objective_SimOpt.hpp:61:3: note: candidate: ROL::ThyraProductME_Objective_SimOpt<Real>::ThyraProductME_Objective_SimOpt(Thyra::ModelEvaluatorDefaultBase<double>&, int, const std::vector<int, std::allocator<int> >&, Teuchos::RCP<Teuchos::ParameterList>, Teuchos::EVerbosityLevel) [with Real = double]
   ThyraProductME_Objective_SimOpt(Thyra::ModelEvaluatorDefaultBase<double>& thyra_model_, int g_index_, const std::vector<int>& p_indices_,
   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/kim/dev/Trilinos_all/Trilinos/packages/piro/src/../../rol/adapters/thyra/src/function/ROL_ThyraProductME_Objective_SimOpt.hpp:61:3: note:   no known conversion for argument 1 from ‘const Thyra::ModelEvaluator<double>’ to ‘Thyra::ModelEvaluatorDefaultBase<double>&’
/home/kim/dev/Trilinos_all/Trilinos/packages/piro/src/../../rol/adapters/thyra/src/function/ROL_ThyraProductME_Objective_SimOpt.hpp:56:7: note: candidate: ROL::ThyraProductME_Objective_SimOpt<double>::ThyraProductME_Objective_SimOpt(const ROL::ThyraProductME_Objective_SimOpt<double>&)
 class ThyraProductME_Objective_SimOpt : public Objective_SimOpt<Real> {
       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/kim/dev/Trilinos_all/Trilinos/packages/piro/src/../../rol/adapters/thyra/src/function/ROL_ThyraProductME_Objective_SimOpt.hpp:56:7: note:   candidate expects 1 argument, 5 provided
/home/kim/dev/Trilinos_all/Trilinos/packages/piro/src/../../rol/adapters/thyra/src/function/ROL_ThyraProductME_Objective_SimOpt.hpp:56:7: note: candidate: ROL::ThyraProductME_Objective_SimOpt<double>::ThyraProductME_Objective_SimOpt(ROL::ThyraProductME_Objective_SimOpt<double>&&)
/home/kim/dev/Trilinos_all/Trilinos/packages/piro/src/../../rol/adapters/thyra/src/function/ROL_ThyraProductME_Objective_SimOpt.hpp:56:7: note:   candidate expects 1 argument, 5 provided
In file included from /home/kim/dev/Trilinos_all/Trilinos/packages/piro/src/Piro_SteadyStateSolver.hpp:159:0,
                 from /home/kim/dev/Trilinos_all/Trilinos/packages/piro/src/Piro_NOXSolver.hpp:46,
                 from /home/kim/dev/Trilinos_all/Trilinos/packages/piro/src/Piro_NOXSolver_Def.hpp:46,
                 from /home/kim/dev/Trilinos_all/Trilinos/packages/piro/src/Piro_NOXSolver.cpp:43:
/home/kim/dev/Trilinos_all/Trilinos/packages/piro/src/Piro_SteadyStateSolver_Def.hpp:432:53: error: no matching function for call to ‘ROL::ThyraProductME_Constraint_SimOpt<double>::ThyraProductME_Constraint_SimOpt(const Thyra::ModelEvaluator<double>&, int&, std::vector<int, std::allocator<int> >&, Teuchos::RCP<Teuchos::ParameterList>&, Teuchos::EVerbosityLevel)’
       ROL::ThyraProductME_Constraint_SimOpt<Scalar> constr(*model_, i, p_indices, opt_paramList, Teuchos::VERB_NONE);
                                                     ^~~~~~
In file included from /home/kim/dev/Trilinos_all/Trilinos/packages/piro/src/Piro_SteadyStateSolver_Def.hpp:80:0,
                 from /home/kim/dev/Trilinos_all/Trilinos/packages/piro/src/Piro_SteadyStateSolver.hpp:159,
                 from /home/kim/dev/Trilinos_all/Trilinos/packages/piro/src/Piro_NOXSolver.hpp:46,
                 from /home/kim/dev/Trilinos_all/Trilinos/packages/piro/src/Piro_NOXSolver_Def.hpp:46,
                 from /home/kim/dev/Trilinos_all/Trilinos/packages/piro/src/Piro_NOXSolver.cpp:43:
/home/kim/dev/Trilinos_all/Trilinos/packages/piro/src/../../rol/adapters/thyra/src/function/ROL_ThyraProductME_Constraint_SimOpt.hpp:65:3: note: candidate: ROL::ThyraProductME_Constraint_SimOpt<Real>::ThyraProductME_Constraint_SimOpt(Thyra::ModelEvaluatorDefaultBase<double>&, int, const std::vector<int, std::allocator<int> >&, Teuchos::RCP<Teuchos::ParameterList>, Teuchos::EVerbosityLevel) [with Real = double]
   ThyraProductME_Constraint_SimOpt(Thyra::ModelEvaluatorDefaultBase<double>& thyra_model_, int g_index_, const std::vector<int>& p_indices_,
   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/kim/dev/Trilinos_all/Trilinos/packages/piro/src/../../rol/adapters/thyra/src/function/ROL_ThyraProductME_Constraint_SimOpt.hpp:65:3: note:   no known conversion for argument 1 from ‘const Thyra::ModelEvaluator<double>’ to ‘Thyra::ModelEvaluatorDefaultBase<double>&’
/home/kim/dev/Trilinos_all/Trilinos/packages/piro/src/../../rol/adapters/thyra/src/function/ROL_ThyraProductME_Constraint_SimOpt.hpp:61:7: note: candidate: ROL::ThyraProductME_Constraint_SimOpt<double>::ThyraProductME_Constraint_SimOpt(const ROL::ThyraProductME_Constraint_SimOpt<double>&)
 class ThyraProductME_Constraint_SimOpt : public Constraint_SimOpt<Real> {
       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/kim/dev/Trilinos_all/Trilinos/packages/piro/src/../../rol/adapters/thyra/src/function/ROL_ThyraProductME_Constraint_SimOpt.hpp:61:7: note:   candidate expects 1 argument, 5 provided
/home/kim/dev/Trilinos_all/Trilinos/packages/piro/src/../../rol/adapters/thyra/src/function/ROL_ThyraProductME_Constraint_SimOpt.hpp:61:7: note: candidate: ROL::ThyraProductME_Constraint_SimOpt<double>::ThyraProductME_Constraint_SimOpt(ROL::ThyraProductME_Constraint_SimOpt<double>&&)
/home/kim/dev/Trilinos_all/Trilinos/packages/piro/src/../../rol/adapters/thyra/src/function/ROL_ThyraProductME_Constraint_SimOpt.hpp:61:7: note:   candidate expects 1 argument, 5 provided
In file included from /home/kim/dev/Trilinos_all/Trilinos/packages/piro/src/Piro_SteadyStateSolver.hpp:159:0,
                 from /home/kim/dev/Trilinos_all/Trilinos/packages/piro/src/Piro_NOXSolver.hpp:46,
                 from /home/kim/dev/Trilinos_all/Trilinos/packages/piro/src/Piro_NOXSolver_Def.hpp:46,
                 from /home/kim/dev/Trilinos_all/Trilinos/packages/piro/src/Piro_NOXSolver.cpp:43:
/home/kim/dev/Trilinos_all/Trilinos/packages/piro/src/Piro_SteadyStateSolver_Def.hpp: In instantiation of ‘void Piro::SteadyStateSolver<Scalar>::evalReducedHessian(const Thyra::ModelEvaluatorBase::InArgs<Scalar>&, const Thyra::ModelEvaluatorBase::OutArgs<Scalar>&) const [with Scalar = double]’:
/home/kim/dev/Trilinos_all/Trilinos/packages/piro/src/Piro_NOXSolver_Def.hpp:205:3:   required from ‘void Piro::NOXSolver<Scalar>::evalModelImpl(const Thyra::ModelEvaluatorBase::InArgs<Scalar>&, const Thyra::ModelEvaluatorBase::OutArgs<Scalar>&) const [with Scalar = double]’
/home/kim/dev/Trilinos_all/Trilinos/packages/piro/src/Piro_NOXSolver.cpp:50:16:   required from here
/home/kim/dev/Trilinos_all/Trilinos/packages/piro/src/Piro_SteadyStateSolver_Def.hpp:1218:50: error: no matching function for call to ‘ROL::ThyraProductME_Objective_SimOpt<double>::ThyraProductME_Objective_SimOpt(const Thyra::ModelEvaluator<double>&, int&, std::vector<int, std::allocator<int> >&, Teuchos::RCP<Teuchos::ParameterList>&, Teuchos::EVerbosityLevel)’
     ROL::ThyraProductME_Objective_SimOpt<Scalar> obj(*model_, g_index, p_indices, opt_paramList, Teuchos::VERB_NONE);
                                                  ^~~
In file included from /home/kim/dev/Trilinos_all/Trilinos/packages/piro/src/Piro_SteadyStateSolver_Def.hpp:79:0,
                 from /home/kim/dev/Trilinos_all/Trilinos/packages/piro/src/Piro_SteadyStateSolver.hpp:159,
                 from /home/kim/dev/Trilinos_all/Trilinos/packages/piro/src/Piro_NOXSolver.hpp:46,
                 from /home/kim/dev/Trilinos_all/Trilinos/packages/piro/src/Piro_NOXSolver_Def.hpp:46,
                 from /home/kim/dev/Trilinos_all/Trilinos/packages/piro/src/Piro_NOXSolver.cpp:43:
/home/kim/dev/Trilinos_all/Trilinos/packages/piro/src/../../rol/adapters/thyra/src/function/ROL_ThyraProductME_Objective_SimOpt.hpp:61:3: note: candidate: ROL::ThyraProductME_Objective_SimOpt<Real>::ThyraProductME_Objective_SimOpt(Thyra::ModelEvaluatorDefaultBase<double>&, int, const std::vector<int, std::allocator<int> >&, Teuchos::RCP<Teuchos::ParameterList>, Teuchos::EVerbosityLevel) [with Real = double]
   ThyraProductME_Objective_SimOpt(Thyra::ModelEvaluatorDefaultBase<double>& thyra_model_, int g_index_, const std::vector<int>& p_indices_,
   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/kim/dev/Trilinos_all/Trilinos/packages/piro/src/../../rol/adapters/thyra/src/function/ROL_ThyraProductME_Objective_SimOpt.hpp:61:3: note:   no known conversion for argument 1 from ‘const Thyra::ModelEvaluator<double>’ to ‘Thyra::ModelEvaluatorDefaultBase<double>&’
/home/kim/dev/Trilinos_all/Trilinos/packages/piro/src/../../rol/adapters/thyra/src/function/ROL_ThyraProductME_Objective_SimOpt.hpp:56:7: note: candidate: ROL::ThyraProductME_Objective_SimOpt<double>::ThyraProductME_Objective_SimOpt(const ROL::ThyraProductME_Objective_SimOpt<double>&)
 class ThyraProductME_Objective_SimOpt : public Objective_SimOpt<Real> {
       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/kim/dev/Trilinos_all/Trilinos/packages/piro/src/../../rol/adapters/thyra/src/function/ROL_ThyraProductME_Objective_SimOpt.hpp:56:7: note:   candidate expects 1 argument, 5 provided
/home/kim/dev/Trilinos_all/Trilinos/packages/piro/src/../../rol/adapters/thyra/src/function/ROL_ThyraProductME_Objective_SimOpt.hpp:56:7: note: candidate: ROL::ThyraProductME_Objective_SimOpt<double>::ThyraProductME_Objective_SimOpt(ROL::ThyraProductME_Objective_SimOpt<double>&&)
/home/kim/dev/Trilinos_all/Trilinos/packages/piro/src/../../rol/adapters/thyra/src/function/ROL_ThyraProductME_Objective_SimOpt.hpp:56:7: note:   candidate expects 1 argument, 5 provided
In file included from /home/kim/dev/Trilinos_all/Trilinos/packages/piro/src/Piro_SteadyStateSolver.hpp:159:0,
                 from /home/kim/dev/Trilinos_all/Trilinos/packages/piro/src/Piro_NOXSolver.hpp:46,
                 from /home/kim/dev/Trilinos_all/Trilinos/packages/piro/src/Piro_NOXSolver_Def.hpp:46,
                 from /home/kim/dev/Trilinos_all/Trilinos/packages/piro/src/Piro_NOXSolver.cpp:43:
/home/kim/dev/Trilinos_all/Trilinos/packages/piro/src/Piro_SteadyStateSolver_Def.hpp:1219:51: error: no matching function for call to ‘ROL::ThyraProductME_Constraint_SimOpt<double>::ThyraProductME_Constraint_SimOpt(const Thyra::ModelEvaluator<double>&, int&, std::vector<int, std::allocator<int> >&, Teuchos::RCP<Teuchos::ParameterList>&, Teuchos::EVerbosityLevel)’
     ROL::ThyraProductME_Constraint_SimOpt<Scalar> constr(*model_, g_index, p_indices, opt_paramList, Teuchos::VERB_NONE);
                                                   ^~~~~~
In file included from /home/kim/dev/Trilinos_all/Trilinos/packages/piro/src/Piro_SteadyStateSolver_Def.hpp:80:0,
                 from /home/kim/dev/Trilinos_all/Trilinos/packages/piro/src/Piro_SteadyStateSolver.hpp:159,
                 from /home/kim/dev/Trilinos_all/Trilinos/packages/piro/src/Piro_NOXSolver.hpp:46,
                 from /home/kim/dev/Trilinos_all/Trilinos/packages/piro/src/Piro_NOXSolver_Def.hpp:46,
                 from /home/kim/dev/Trilinos_all/Trilinos/packages/piro/src/Piro_NOXSolver.cpp:43:
/home/kim/dev/Trilinos_all/Trilinos/packages/piro/src/../../rol/adapters/thyra/src/function/ROL_ThyraProductME_Constraint_SimOpt.hpp:65:3: note: candidate: ROL::ThyraProductME_Constraint_SimOpt<Real>::ThyraProductME_Constraint_SimOpt(Thyra::ModelEvaluatorDefaultBase<double>&, int, const std::vector<int, std::allocator<int> >&, Teuchos::RCP<Teuchos::ParameterList>, Teuchos::EVerbosityLevel) [with Real = double]
   ThyraProductME_Constraint_SimOpt(Thyra::ModelEvaluatorDefaultBase<double>& thyra_model_, int g_index_, const std::vector<int>& p_indices_,
   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/kim/dev/Trilinos_all/Trilinos/packages/piro/src/../../rol/adapters/thyra/src/function/ROL_ThyraProductME_Constraint_SimOpt.hpp:65:3: note:   no known conversion for argument 1 from ‘const Thyra::ModelEvaluator<double>’ to ‘Thyra::ModelEvaluatorDefaultBase<double>&’
/home/kim/dev/Trilinos_all/Trilinos/packages/piro/src/../../rol/adapters/thyra/src/function/ROL_ThyraProductME_Constraint_SimOpt.hpp:61:7: note: candidate: ROL::ThyraProductME_Constraint_SimOpt<double>::ThyraProductME_Constraint_SimOpt(const ROL::ThyraProductME_Constraint_SimOpt<double>&)
 class ThyraProductME_Constraint_SimOpt : public Constraint_SimOpt<Real> {
       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/kim/dev/Trilinos_all/Trilinos/packages/piro/src/../../rol/adapters/thyra/src/function/ROL_ThyraProductME_Constraint_SimOpt.hpp:61:7: note:   candidate expects 1 argument, 5 provided
/home/kim/dev/Trilinos_all/Trilinos/packages/piro/src/../../rol/adapters/thyra/src/function/ROL_ThyraProductME_Constraint_SimOpt.hpp:61:7: note: candidate: ROL::ThyraProductME_Constraint_SimOpt<double>::ThyraProductME_Constraint_SimOpt(ROL::ThyraProductME_Constraint_SimOpt<double>&&)
/home/kim/dev/Trilinos_all/Trilinos/packages/piro/src/../../rol/adapters/thyra/src/function/ROL_ThyraProductME_Constraint_SimOpt.hpp:61:7: note:   candidate expects 1 argument, 5 provided
packages/piro/src/CMakeFiles/piro.dir/build.make:206: recipe for target 'packages/piro/src/CMakeFiles/piro.dir/Piro_NOXSolver.cpp.o' failed
make[2]: *** [packages/piro/src/CMakeFiles/piro.dir/Piro_NOXSolver.cpp.o] Error 1
CMakeFiles/Makefile2:76691: recipe for target 'packages/piro/src/CMakeFiles/piro.dir/all' failed
make[1]: *** [packages/piro/src/CMakeFiles/piro.dir/all] Error 2
Makefile:162: recipe for target 'all' failed
make: *** [all] Error 2

```

I think that those modifications are fine; they increase the range of applicability of those 3 objects.